Improve file migration information to documentation (#9026)
silverstripe/silverstripe-framework: docs/en/02_Developer_Guides/14_Files/05_File_Migration.md

## Natural path vs Hash Path

Two types of paths are used by SilverStripe CMS when serving files:
* Natural paths and
* Hash paths.

Natural paths will match the path displayed in SilverStripe's asset management section (e.g.: `assets/Folder/form.pdf`). Hash paths will include a partial hash as directory name (e.g.: `assets/Folder/282318025b/form.pdf`). The hash is generated by computing the sha1 hash of the provided file and keeping the first ten characters.

The hash path has the advantage of being unique for every version of the file your users might upload. This allows multiple versions of the same file to co-exist, enabling features like [file archives](04_File_Storage#archived). This is why hash path is the default file format for the protected file store.

The Natural path has the advantage of being easier to understand and permanent. When a newer version of a file is published to a natural path, it will be accessible at the same URL as the previous version. This is why natural path is the default file format for the public file store.

> [Learn more about SilverStripe File Store](04_File_Storage)

Author: ```suggestion
[Learn more about SilverStripe File Storage](04_File_Storage)
```

## Migration from SilverStripe 3 to Silverstripe 4.4 or later

> While the SilverStripe 3 file paths may look similar to the natural paths in SilverStripe 4.4, the database entries for each file have to be normalise to conform to the new SilverStripe 4 structure. File variants (e.g.: image thumbnails) are stored differently in SilverStripe 4. They will automatically be moved to their new expected location.

Author: ```suggestion
While the SilverStripe 3 file paths may look similar to the natural paths in SilverStripe 4.4, the database entries for each file have to be normalised to conform to the new SilverStripe 4 structure. File variants (e.g. image thumbnails) are stored differently in SilverStripe 4. They will automatically be moved to their new location.
```

> SilverStripe 3 did not support draft file. So all migrated SilverStripe 3 files will automatically be published by the File Migration Task.

Author: ```suggestion
SilverStripe 3 did not support draft files which means that all migrated SilverStripe 3 files will automatically be published by the File Migration Task.
```

> SilverStripe 4 has different naming convention. File names can not contain "double underscores" in SilverStripe 4 (`__`) because those characters are used to indicate a "file variant". SilverStripe 3 files not conforming to this rule will automatically be renamed by the file migration task.

Author: ```suggestion
File names can not contain "double underscores" in SilverStripe 4 (`__`) because those characters are used to indicate a "file variant". SilverStripe 3 files not conforming to this rule will automatically be renamed by the file migration task.
```

> SilverStripe 3 files not conforming to this rule will automatically be renamed by the file migration task.

To what?

Reviewer: I added a complement of info

### Will old URLs to SilverStripe 3 files still resolve?

> In most cases, the SilverStripe 3 file path will be identical to the public path of the file in SilverStripe 4. Links to SilverStripe 3 file variants (e.g.: cropped images) will automatically be redirected to their new location.

Author: ```suggestion
In most cases, the SilverStripe 3 file path will be identical to the public path of the file in SilverStripe 4. Links to SilverStripe 3 file variants (e.g. cropped images) will automatically be redirected to their new location.
```

### What about file references in `HTMLText` database fields?

> The short code format in SilverStripe 4 has changed from SilverStripe 3. This means that when CMS editors try to edit existing content and alter image or file links, the reference to the file will not work.

Author: ```suggestion
The shortcode format in SilverStripe 4 has changed from SilverStripe 3. This means that when CMS editors try to edit existing content and alter image or file links, the reference to the file will not work.
```

> After running the file migration, you can run a the short code migration task to update your existing HTML content in your database.

Author: ```suggestion
After running the file migration, you can run the short code migration task to update your existing HTML content in your database:

```bash
sake dev/tasks/TagsToShortcode
``` 

> This will rewrite your existing shortcode to the newer format SilverStripe 4 expects.

Author: ```suggestion
This will rewrite your existing shortcodes to the newer format SilverStripe 4 expects as well as convert `img` and `a` tags to use shortcodes..
```

## Migrating from SilverStripe 4.3 or below to SilverStripe 4.4 or above

Silverstripe 4.0 to 4.3 were using hash paths for the public file store as well as the protected file store. This meant that every time a new version of a file was published, it would be served under a different URL.

SilverStripe 4.4 and later default to using natural paths for public store files. After upgrading to SilverStripe 4.4 or above, you can run the file migration task to normalise your existing files.

This will take all existing files under a hash path and move them to their natural path.

### Do I need to re-migrate my files?

No, you do not need to run the file migration task when upgrading to SilverStripe 4.4 or later. 

Your existing files will still resolve. Old hash paths will be redirected to the latest version of the file. As newer versions of existing files are published, they will be moved to their natural path.

You can choose to run the migration task at any point.

### What if I’m using `legacy_filenames`?

> SilverStripe 4.0 to 4.3 allowed you to enable a `legacy_filenames` mode. This would be equivalent to using natural path for both the public store and the protected store.

Author: ```suggestion
SilverStripe 4.0 to 4.3 allowed you to enable a `legacy_filenames` mode. This would be equivalent to using natural paths for both the public store and the protected store.
```

> With the introduction of natural path for the public store, the primary reason for `legacy_filenames` mode is no longer relevant. `legacy_filenames` mode is now considered deprecated, however leaving it enabled will not have any impact on your project – it will simply be ignored.

Author: ```suggestion
With the introduction of natural paths for the public store, the primary reason for `legacy_filenames` mode is no longer relevant. `legacy_filenames` mode is now considered deprecated, however leaving it enabled will not have any impact on your project – it will simply be ignored.
```

###  What if I want to keep using hash paths for my public files?

If you want to keep using hash paths for your public store, you can add the following entries to your YML configuration files.

```yml
SilverStripe\Core\Injector\Injector:
  # Define public resolution strategy
  SilverStripe\Assets\FilenameParsing\FileResolutionStrategy.public:
    class: SilverStripe\Assets\FilenameParsing\FileIDHelperResolutionStrategy
    properties:
    DefaultFileIDHelper: '%$SilverStripe\Assets\FilenameParsing\HashFileIDHelper'
    ResolutionFileIDHelpers:
      - '%$SilverStripe\Assets\FilenameParsing\HashFileIDHelper'
      VersionedStage: Live
  # Define protected resolution strategy
  SilverStripe\Assets\FilenameParsing\FileResolutionStrategy.protected:
    class: SilverStripe\Assets\FilenameParsing\FileIDHelperResolutionStrategy
    properties:
      DefaultFileIDHelper: '%$SilverStripe\Assets\FilenameParsing\HashFileIDHelper'
      ResolutionFileIDHelpers:
        - '%$SilverStripe\Assets\FilenameParsing\HashFileIDHelper'
      VersionedStage: Stage
```

## Disabling old file resolution format

> When starting a brand new project, or after you’ve run the file migration task, all your files will be in their default locations. This gives you the option to disable alternative resolution formats. This will slightly speed up file resolution.

Author: ```suggestion
When starting a brand new project, or after you’ve ran the file migration task, all your files will be in their default locations. This gives you the option to disable alternative resolution formats. This will slightly speed up file resolution.
```

SilverStripe\Dev\Tasks\MigrateFileTask:
  extensions:
    - MigrateFileTaskExtension
=======

> Your exact experience will vary based on your host server, the size of your files and other conditions. If your site is hosted on a managed environment (e.g.: [Common Web Platform](https://www.cwp.govt.nz/service-desk) or [SilverStripe Platform](https://docs.platform.silverstripe.com/support/)), you may not have access to the command line to manually run the migration task. Contact your hosting provider's helpdesk if that's your case.

Author: ```suggestion
Your exact experience will vary based on your host server, the size of your files and other conditions. If your site is hosted on a managed environment (e.g. [Common Web Platform](https://www.cwp.govt.nz/service-desk) or [SilverStripe Platform](https://docs.platform.silverstripe.com/support/)), you may not have access to the command line to manually run the migration task. Contact your hosting provider's helpdesk if that's your case.
```
